### ISSUE TITLE
Sustainable exports

### ISSUE THREAD

**Issue by sneridagh** [plone/plone.exportimport#39]
If we want to use `plone.exportimport` not just for "snapshots"/migrations but also for maintaining example content, we need to fix these:

- It overrides always `modified` and `created` which are not needed/relevant, maybe for a migration, but not for a snapshot/example content.
- The `exportimport.versions` field contain an object that changes the UID always... probably to the "previous" version of the last export... 
- It also exports fields with no value to `null`. In Plone RESTAPI idiom this is not needed, since in import all non existent fields should remain empty, not set.
- It exports `parents` and `items` this are not required, nor needed by the import.

Maybe we should differentiate both use cases (snapshots/migration and example content)?

**Comment by mauritsvanrees** [comment]
I am looking at some of this, and also issue #36.

> If we want to use `plone.exportimport` not just for "snapshots"/migrations but also for maintaining example content, we need to fix these:
> 
> * It overrides always `modified` and `created` which are not needed/relevant, maybe for a migration, but not for a snapshot/example content.

I agree that for demo content it could be good to ignore these.  For example on https://classic.demo.plone.org when you login you see that the last modification date of the front page is 8 March 2024.  For a demo site this could simply be the actual modification date, so a few fours ago.

> * The `exportimport.versions` field contain an object that changes the UID always... probably to the "previous" version of the last export...

Correct.  In the importer we [get or create the object instance](https://github.com/plone/plone.exportimport/blob/main/src/plone/exportimport/importers/content.py#L74) and this already creates a version in the history, with an object that has a random UID and no field data yet, not even a title.

It might be possible to prevent creating a version history entry for this initial bare new content item.  Maybe temporarily remove `plone.versioning` from the behaviors and restore it after the instance has been created.  Or remove the type from the versional content type in the `portal_repository`.

But it seems easier to completely skip exporting the versions.  This could be an option in the exporter: add `--include-revisions` to the cli, and pass this option on to the exporters.  I have some work-in-progress code for this.

> * It also exports fields with no value to `null`. In Plone RESTAPI idiom this is not needed, since in import all non existent fields should remain empty, not set.

I am not sure about this, but I will take your word for it.

In [`cleanup_export_data`](https://github.com/plone/plone.exportimport/blob/1.0.0a8/src/plone/exportimport/utils/content/export_helpers.py#L51) we could drop such fields.

> * It exports `parents` and `items` this are not required, nor needed by the import.

In [`cleanup_export_data`](https://github.com/plone/plone.exportimport/blob/1.0.0a8/src/plone/exportimport/utils/content/export_helpers.py#L51) we are already dropping the `items` key.

I think we could indeed also drop the `parent` key, except that this needs a change in [`get_parent_from_item`](https://github.com/plone/plone.exportimport/blob/1.0.0a8/src/plone/exportimport/utils/content/import_helpers.py#L45-L46), otherwise no container is found, and the item cannot be added.  I have some code for that.

> Maybe we should differentiate both use cases (snapshots/migration and example content)?

We could differentiate by adding command line options to the exporters in both `plone.distribution` and `plone.exportimport`.  This may need changes in all individual exporters.  I am looking into it.

Question is what the defaults would be.  Do we export something by default, and allow disabling this?  Or do we not export it, and allow enabling it?  I would say by default:

* Do not export `created/modified` dates.  Maybe the same for published/expired dates, but I did not look at those yet.  Enable with `--include-dates`.
* Do not export versions/revisions.  Enable with `--include-revisions`.
* Do not export fields with value None.  No option.
* Do not export parent.  No option.

What do you think?

**Comment by mauritsvanrees** [comment]
For that third use case, you can probably manually edit the files and remove the uid, or monkey patch [`cleanup_export_data`](https://github.com/plone/plone.exportimport/blob/82a9524d13ec711ac00eb54d30f8f97e5b9c7706/src/plone/exportimport/utils/content/export_helpers.py#L51) to also drop the `uid`.  Or patch to use an [extra cleaner method](https://github.com/plone/plone.exportimport/blob/main/src/plone/exportimport/utils/content/export_helpers.py#L331).

Otherwise I am inclined to say: this is not what `plone.exportimport` is meant for, please try adding an option in `collective.exportimport` instead.

**Comment by sneridagh** [comment]
@mauritsvanrees all the options you mentioned look good to me and would add lots of value!

**Comment by yurj** [comment]
> For that third use case, you can probably manually edit the files and remove the uid, or monkey patch [`cleanup_export_data`](https://github.com/plone/plone.exportimport/blob/82a9524d13ec711ac00eb54d30f8f97e5b9c7706/src/plone/exportimport/utils/content/export_helpers.py#L51) to also drop the `uid`. Or patch to use an [extra cleaner method](https://github.com/plone/plone.exportimport/blob/main/src/plone/exportimport/utils/content/export_helpers.py#L331).
> 
> Otherwise I am inclined to say: this is not what `plone.exportimport` is meant for, please try adding an option in `collective.exportimport` instead.

yes, now I just remove the uids, relations and links works because import look at the paths I suppose.

**Comment by davisagli** [comment]
@mauritsvanrees 

> Do not export created/modified dates. Maybe the same for published/expired dates, but I did not look at those yet. Enable with --include-dates.

Instead of excluding these from the export, another option would be to make sure that they are set correctly from the file during an import. Then they would not change needlessly, but would keep a record of when the item was created/modified in any site.

> Do not export versions/revisions. Enable with --include-revisions.

That makes sense to me as a default.

> Do not export fields with value None. No option.

This could be a problem in an edge case where there is a field that has a default value but can optionally be set to None.

> Do not export parent. No option.

Isn't this how the importer knows where to add the item? https://github.com/plone/plone.exportimport/blob/main/src/plone/exportimport/utils/content/import_helpers.py#L32

**Comment by mauritsvanrees** [comment]
About created/modified dates:

> Instead of excluding these from the export, another option would be to make sure that they are set correctly from the file during an import. Then they would not change needlessly, but would keep a record of when the item was created/modified in any site.

That was my original plan.  Done as part of PR #40.

But additionally it could be useful to optionally not export these.

> > Do not export fields with value None. No option.
> 
> This could be a problem in an edge case where there is a field that has a default value but can optionally be set to None.

Okay, then let's leave this alone.

> > Do not export parent. No option.
> 
> How will the importer know where to add the item without this?

See PR #40.  Basically, we get the parent with:

```
parent_path = str(Path(data["@id"]).parent)
parent = api.content.get(path=parent_path)
```

**Comment by mauritsvanrees** [comment]
I have closed my original PR, and instead created four smaller PRs.  That way we can merge obviously wanted changes, without being held up by changes that are more adventurous and where we may have different opinions.

* https://github.com/plone/plone.exportimport/pull/41
* https://github.com/plone/plone.exportimport/pull/42
* https://github.com/plone/plone.exportimport/pull/43
* https://github.com/plone/plone.exportimport/pull/44

**Comment by sneridagh** [comment]
Can't wait for all these to be merged and released! Thanks @mauritsvanrees for working on them!

**Comment by mauritsvanrees** [comment]
I have released `plone.exportimport` 1.0.0b1 with all the changes.

**Comment by mauritsvanrees** [comment]
To support the `--include-revisions` argument in `bin/export-distribution` as well, we need this PR: https://github.com/plone/plone.distribution/pull/108
